Hmr and proxy interfere when proxying websockets requests

Hi,
When using the proxy feature to forward a websockets request to an upstream server, the hmr implementation throws an error and crashes the `dev` command.

It appears that now the hmr module and the proxy module register both handlers for the 'upgrade' event sharing the same http server. The hmr module expects the data of the 'message' event to be JSON and throws if it cannot deserialize the payload. The hrm 'upgrade' handler fires although the request is meant to be processed by the proxy features.

Previously I was proxying request to signalr and the hmr module didn't interfere as it was registered as a standalone websockets server on port 12321.

Thanks!

Germán

```
  [log] Upgrading to WebSocket

undefined:1
{"protocol":"json","version":1}▲
                               ^
SyntaxError: Unexpected token ▲ in JSON at position 31
    at JSON.parse (<anonymous>)
    at WebSocket.<anonymous> (C:\dev\Sonica\ClientApp\node_modules\snowpack\dist-node\index.js:2775:28)
    at WebSocket.emit (events.js:315:20)
    at Receiver.receiverOnMessage (C:\dev\Sonica\ClientApp\node_modules\snowpack\node_modules\ws\lib\websocket.js:800:20)
    at Receiver.emit (events.js:315:20)
    at Receiver.dataMessage (C:\dev\Sonica\ClientApp\node_modules\snowpack\node_modules\ws\lib\receiver.js:436:14)
    at Receiver.getData (C:\dev\Sonica\ClientApp\node_modules\snowpack\node_modules\ws\lib\receiver.js:366:17)
    at Receiver.startLoop (C:\dev\Sonica\ClientApp\node_modules\snowpack\node_modules\ws\lib\receiver.js:142:22)
    at Receiver._write (C:\dev\Sonica\ClientApp\node_modules\snowpack\node_modules\ws\lib\receiver.js:77:10)
    at doWrite (_stream_writable.js:385:12)
    at writeOrBuffer (_stream_writable.js:367:5)
    at Receiver.Writable.write (_stream_writable.js:307:12)
    at Socket.socketOnData (C:\dev\Sonica\ClientApp\node_modules\snowpack\node_modules\ws\lib\websocket.js:875:35)
    at Socket.emit (events.js:315:20)
    at addChunk (_stream_readable.js:296:12)
    at readableAddChunk (_stream_readable.js:272:9)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @ start: `snowpack dev`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the @ start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\German\AppData\Roaming\npm-cache\_logs\2020-06-02T15_36_35_001Z-debug.log
PS C:\dev\Sonica\ClientApp> 
```

--- a/assets/hmr.js
+++ b/assets/hmr.js
@@ -22,7 +22,7 @@ function sendSocketMessage(msg) {
 const socketURL =
   window.HMR_WEBSOCKET_URL ||
   (location.protocol === 'http:' ? 'ws://' : 'wss://') + location.host + '/';
-const socket = new WebSocket(socketURL);
+const socket = new WebSocket(socketURL, 'snowpack-hmr');
 socket.addEventListener('open', () => {
   SOCKET_MESSAGE_QUEUE.forEach(_sendSocketMessage);
   SOCKET_MESSAGE_QUEUE = [];

--- a/src/hmr-server-engine.ts
+++ b/src/hmr-server-engine.ts
@@ -19,9 +19,12 @@ export class EsmHmrEngine {
       : new WebSocket.Server({port: 12321});
     if (options.server) {
       options.server.on('upgrade', (req, socket, head) => {
-        wss.handleUpgrade(req, socket, head, (client) => {
-          wss.emit('connection', client, req);
-        });
+        let isHmrRequest = req.headers['sec-websocket-protocol'] === 'snowpack-hmr';
+        if (isHmrRequest) {
+          wss.handleUpgrade(req, socket, head, (client) => {
+            wss.emit('connection', client, req);
+          });          
+        }
       });
     }
     wss.on('connection', (client) => {

